ambry-api/src/main/java/com/github/ambry/server/ServerSecurityService.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.github.ambry.server;
+
+import com.github.ambry.rest.RestRequest;
+import com.github.ambry.router.Callback;
+import com.github.ambry.router.FutureResult;
+import java.io.Closeable;
+import io.netty.channel.ChannelHandlerContext;
+import java.util.concurrent.Future;
+
+
+/**
+ * Responsible for performing any security validations on the HTTP2 connection terminating on server. Validations
+ * could involve authentication, authorization, security checks and so on which the implementation can decide.
+ * This could also involve setting headers while responding, based on the request.
+ * Exceptions are returned via {@link Callback}s on any validation failure.
+ */
+public interface ServerSecurityService extends Closeable {
+
+  /**
+   * Performs security validations (if any) before allowing the HTTP2 connection setup to be complete and invokes the
+   * {@code callback} once done.
+   * @param ctx the {@link ChannelHandlerContext} to process.
+   * @param callback the callback to invoke once processing is finished.
+   */
+  void validateConnection(ChannelHandlerContext ctx, Callback<Void> callback);
```

Comment:
This interface looks a little netty focused and requires a new dependency for ambry-api. Could this method just take in the `SSLSession` object itself?

Additionally, I wanted to learn a bit more about the motivation for this method as opposed to doing all security validation at request time. Is this an optimization so that SSL related checks only have to be checked once since the peer certificate is the same for all requests on a single connection?

Reply:
1. Thanks @cgtz for the idea. agree it is netty focused. I have changed the interface to accept SSLSession instead without loosing any information. 
2. Yes, as mentioned in the document, currently I am planning to do the security validation on the certificate provided during the HTTP2 SSL connection setup itself. There is no separate certs/SSL negotiation for individual request stream inside the HTTP2 connection. So There is no need to do the cert validation on per request basis. I have added the validateRequest() interface now as well to do any extra security checks for individual request that we may want to do in future.